### DISCUSSION TITLE
API Route works locally, but not in preview mode deployed on Vercel

### DISCUSSION THREAD

**Discussion by jordanjustice** [vercel/next.js#22308] [Help]
I've created an API route for a simple mailchimp email signup, modeled after @leerob's great [post](https://leerob.io/blog/mailchimp-next-js).

It works locally, but the function fails on the preview when deployed to Vercel. I've logged the env's to verify that they exist on the preview mode, and they in fact do.

Any ideas on what else might be happening between the two environments to make them behave differently?

**Comment by leerob** [comment]
1. I'd confirm it works locally with a production build (`next build && next start`)
2. Then confirm the environment variables (which it sounds like you have)
3. Add logging to see where in the function it's failing

**Reply by jordanjustice** [reply]
Thanks for the quick reply there!
1. Works locally in production build.
2. Yep!
3. Added some more logging and see I'm getting a 401.

So I tried it in Postman, and there are no auth issues there.

**Reply by leerob** [reply]
Can you share your code?

**Reply by jordanjustice** [reply]
Sure thing.

```js
// pages/api/subscribe.js
import fetch from "isomorphic-unfetch";

export default async (req, res) => {
  const { email } = req.body;

  if (!email) {
    return res.status(400).json({ error: "Email is required" });
  }

  try {
    const LIST_ID = process.env.MAILCHIMP_LIST_ID;
    const API_KEY = process.env.MAILCHIMP_API_KEY;
    //API keys are in the form <key>-us3.
    const DATACENTER = API_KEY.split("-")[1];

    // The status of 'subscribed' is equivalent to a double opt-in.
    const data = {
      email_address: email,
      status: "subscribed",
    };

    const response = await fetch(
      `https://${DATACENTER}.api.mailchimp.com/3.0/lists/${LIST_ID}/members`,
      {
        body: JSON.stringify(data),
        headers: {
          Authorization: `apikey ${API_KEY}`,
          "Content-Type": "application/json",
        },
        method: "POST",
      }
    );

    console.log("response :>> ", response);

    if (response.status >= 400) {
      return res.status(400).json({
        error: `❗ Oops, we had trouble adding you to the list. Send us an email at [subscribe@toledosunday.com] and we'll add you to the list.`,
      });
    }

    return res.status(201).json({ error: "" });
  } catch (error) {
    return res.status(500).json({ error: error.message || error.toString() });
  }
};

```

**Reply by leerob** [reply]
Looks like they might have changed their auth format: https://mailchimp.com/developer/marketing/guides/create-your-first-audience/#add-a-contact-to-an-audience

Could you try Authorization: `Basic ${API_KEY}`?

**Reply by jordanjustice** [reply]
Still no luck with the following (and still works locally with a prod build):
```js
const response = await fetch(
      `https://${DATACENTER}.api.mailchimp.com/3.0/lists/${LIST_ID}/members`,
      {
        body: JSON.stringify(data),
        headers: {
          Authorization: `Basic ${API_KEY}`,
          "Content-Type": "application/json",
        },
        method: "POST",
      }
    );
    ```

**Reply by jordanjustice** [reply]
Great apologies... I had one character trimmed off my API key. 🤦🏼
